lib/extensions/ar_migration.rb

```diff
@@ -1,6 +1,14 @@
 module ArPglogicalMigrationHelper
-  def self.discover_schema_migrations_ran_class
-    return unless ActiveRecord::Base.connection.table_exists?("schema_migrations_ran")
+  def self.discover_schema_migrations_ran_class(version)
+    # Schema versions less than 20171031010000 certainly don't have the table, so we can
+    # avoid excessive queries but since this method is called both before AND after a migration
+    # from the ArPglogicalMigration prepended module, 20171031010000 is different based on direction:
+    #   migrate up   - before: missing, after: exists
+    #   migrate down - before: exists,  after: missing
+    # Therefore, we need to query the table for that migration.
+    return if version < "20171031010000"
+    return if version == "20171031010000" && !ActiveRecord::Base.connection.table_exists?("schema_migrations_ran")
```

Comment:
@Fryguy I think this logic finally makes sense.  We only need to query for the tables existence for the migration that added it since it may or may not exist before/after the migration (this is called from both) depending on the direction (up or down).  Before the migration, we can return early.  Versions after the migration, we don't need to check because the table exists.